src/ClassDiscovery.php

```diff
@@ -2,126 +2,104 @@
+use Http\Discovery\Exception\StrategyUnavailableException;
 
 /**
  * Registry that based find results on class existence.
  *
  * @author David de Boer <david@ddeboer.nl>
  * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
  */
 abstract class ClassDiscovery
 {
     /**
-     * @var GeneratedPuliFactory
+     * A list of strategies to find classes.
+     *
+     * @var array
      */
-    private static $puliFactory;
+    public static $strategies = [
+        Strategy\Puli::class,
+        Strategy\HttpClients::class,
+        Strategy\GuzzleFactory::class,
+        Strategy\DiactorosFactory::class,
+    ];
 
     /**
-     * @var Discovery
+     * Discovery cache to make the second time we use discovery faster.
+     *
+     * @var array
      */
-    private static $puliDiscovery;
+    private static $cache = [];
 
     /**
-     * @return GeneratedPuliFactory
-     */
-    public static function getPuliFactory()
-    {
-        if (null === self::$puliFactory) {
-            if (!defined('PULI_FACTORY_CLASS')) {
-                throw new \RuntimeException('Puli Factory is not available');
-            }
-
-            $puliFactoryClass = PULI_FACTORY_CLASS;
-
-            if (!class_exists($puliFactoryClass)) {
-                throw new \RuntimeException('Puli Factory class does not exist');
-            }
-
-            self::$puliFactory = new $puliFactoryClass();
-        }
-
-        return self::$puliFactory;
-    }
-
-    /**
-     * Sets the Puli factory.
+     * Finds a class.
      *
-     * @param object $puliFactory
+     * @param string $type
+     *
+     * @return string
+     *
+     * @throws DiscoveryFailedException
      */
-    public static function setPuliFactory($puliFactory)
+    protected static function findOneByType($type)
     {
-        if (!is_callable([$puliFactory, 'createRepository']) || !is_callable([$puliFactory, 'createDiscovery'])) {
-            throw new \InvalidArgumentException('The Puli Factory must expose a repository and a discovery');
+        // Look in the cache
+        if (null !== $class = self::getFromCache($type)) {
+            return $class;
         }
 
-        self::$puliFactory = $puliFactory;
-        self::$puliDiscovery = null;
-    }
+        $exceptions = [];
+        foreach (self::$strategies as $strategy) {
+            try {
+                $bindings = call_user_func($strategy.'::find', $type);
+            } catch (StrategyUnavailableException $e) {
+                $exceptions[] = $e;
+                continue;
+            }
 
-    /**
-     * Resets the factory.
-     */
-    public static function resetPuliFactory()
-    {
-        self::$puliFactory = null;
-        self::$puliDiscovery = null;
-    }
+            foreach ($bindings as $binding) {
+                if ($binding['condition']) {
+                    if (!self::evaluateCondition($binding['condition'])) {
```

Comment:
I assumed (maybe wrongly) that they are static. 

The current cache implementation is only valid for a request and our conditions are checking if the class exists. But since we support conditions being a callable we may need to rerun the conditions to be sure.